pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,290 @@
+	}
+	return "apis/" + gvk.GroupVersion().String()
+}
+
+// parserFnForSchemer returns a BuildParserFn that builds a new *GvkParser
+// from the OpenAPI GroupVersion schema client
+func parserFnForSchemer(schemer *CustomOAPIGroupVersion) func() (*GvkParser, error) {
+	return func() (*GvkParser, error) {
+		s, err := schemer.Schema("application/json")
+		if err != nil {
+			return nil, err
+		}
+		var oapi spec3.OpenAPI
+		if err := json.Unmarshal(s, &oapi); err != nil {
+			return nil, err
+		}
+
+		specs := map[string]*spec.Schema{}
+		for k, v := range oapi.Components.Schemas {
+			specs[k] = v
+		}
+		return NewGVKParser(specs, false)
+	}
+}
+
+func (e *cachingUnstructuredExtractor) objectForGVK(gvk schema.GroupVersionKind) (*typed.ParseableType, error) {
+	schemer, err := e.schemerForGVK(gvk)
+	if err != nil {
+		return nil, err
+	}
+
+	parser, err := e.cache.ParserForGV(gvk.GroupVersion().String(), schemer.etag, parserFnForSchemer(schemer))
+	if err != nil {
+		return nil, err
+	}
+	return parser.Type(gvk), nil
+}
+
+func (e *cachingUnstructuredExtractor) extractUnstructured(object *unstructured.Unstructured, fieldManager string, subresource string) (*unstructured.Unstructured, error) {
+	objectType, err := e.objectForGVK(object.GroupVersionKind())
+	if err != nil {
+		return nil, err
+	}
+	result := &unstructured.Unstructured{}
+
+	err = managedfields.ExtractInto(object, *objectType, fieldManager, result, subresource) //nolint:forbidigo
+	if err != nil {
+		return nil, errors.Wrap(err, "failed calling ExtractInto for unstructured")
+	}
+	result.SetName(object.GetName())
+	result.SetNamespace(object.GetNamespace())
+	result.SetKind(object.GetKind())
+	result.SetAPIVersion(object.GetAPIVersion())
+	return result, nil
+}
+
+// CustomOAPIGroupVersion is the customized variant of the unexported
+// openapi.groupversion with ETag extracted.
+// see https://github.com/kubernetes/client-go/blob/78c1586020d8bef4d031a556f867544ca34845a1/openapi/groupversion.go#L32
+type CustomOAPIGroupVersion struct {
+	restClient      rest.Interface
+	item            handler3.OpenAPIV3DiscoveryGroupVersion
+	useClientPrefix bool
+	etag            string
+}
+
+// NewCustomOAPIGroupVersion returns a new CustomOAPIGroupVersion instance
+func NewCustomOAPIGroupVersion(client rest.Interface, item handler3.OpenAPIV3DiscoveryGroupVersion, useClientPrefix bool, etag string) *CustomOAPIGroupVersion {
+	return &CustomOAPIGroupVersion{restClient: client, item: item, useClientPrefix: useClientPrefix, etag: etag}
+}
+
+// Schema returns the OpenAPI schema for the OpenAPI GroupVersion
+// adapted from
+// https://github.com/kubernetes/client-go/blob/78c1586020d8bef4d031a556f867544ca34845a1/openapi/groupversion.go#L42
+func (g *CustomOAPIGroupVersion) Schema(contentType string) ([]byte, error) {
+	if !g.useClientPrefix {
+		return g.restClient.Get().
+			RequestURI(g.item.ServerRelativeURL).
+			SetHeader("Accept", contentType).
+			Do(context.TODO()).
```

Comment:
for this particular function `Schema()` , adding `ctx` would change the interface `openapi.GroupVersion`. We anyways directly use the struct and do not make use of the interface , so I think it is OK 👍

Reply:
yeah indeed it would, I think it's ok to not conform with the schema for now, adding ctx in our "fork" which is hopefully temporary, I think you should propose adding ctx upstream

Reply:
done